### PR TITLE
[ONNX] Enable _jit_pass_onnx_fold_if only when dynamic_axes is None

### PR DESCRIPTION
**Pull request by jiafatom**
Fixing https://github.com/pytorch/vision/issues/3251 (PR https://github.com/pytorch/pytorch/pull/49410 triggers the torch vision test build failure, on three tests test_faster_rcnn, test_mask_rcnn, test_keypoint_rcnn. )

The offending PR is fine on pytorch UT, because the torchvision and pytorch test has a gap when we merge them - we are using different test API on two sides, therefore causing some discrepancy.

This PR bridge the gap for the above three tests, and ~~disable `_jit_pass_onnx_fold_if` pass until it gets fixed.~~ 
Allow `_jit_pass_onnx_fold_if` only when `dynamic_axes is None`.

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -209,7 +209,8 @@ def _optimize_graph(graph, operator_export_type, _disable_torch_constant_prop=Fa
         torch._C._jit_pass_onnx_scalar_type_analysis(graph)
         torch._C._jit_pass_lint(graph)
 
-        torch._C._jit_pass_onnx_fold_if(graph)
+        if dynamic_axes is None or not bool(dynamic_axes):
```
Comment (BowenBao) [inline torch/onnx/utils.py:212]: `not bool(dynamic_axes)` this looks strange.. which exact case is this guarding against?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -410,8 +410,13 @@ def test_faster_rcnn(self):
...
         self.run_test(model, (x,), input_names=["images_tensors"], output_names=["outputs"],
                       dynamic_axes={"images_tensors": [0, 1, 2, 3], "outputs": [0, 1, 2, 3]}, rtol=1e-3, atol=1e-5)
+        dummy_image = [torch.ones(3, 100, 100) * 0.3]
+        images, test_images = self.get_test_images()
+        self.run_test(model, (images,), test_with_inputs=[(images,), (test_images,), (dummy_image,)],
+                      input_names=["images_tensors"], output_names=["outputs"],
+                      dynamic_axes={"images_tensors": [0, 1, 2], "outputs": [0, 1, 2]}, rtol=1e-3, atol=1e-5)
 
-    def get_image_from_url(self, url):
+    def get_image_from_url(self, url, size=None):
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:419]: nit: size=(300, 200) and remove below?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -426,15 +431,24 @@ def get_image_from_url(self, url):
...
-        image = self.get_image_from_url(url=image_url)
+        image = self.get_image_from_url(url=image_url, size=(100, 320))
+
+        image_url2 = "https://pytorch.org/tutorials/_static/img/tv_tutorial/tv_image05.png"
+        image2 = self.get_image_from_url(url=image_url2, size=(250, 380))
+
         images = [image]
-        return images
+        test_images = [image2]
+        return images, test_images
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:451]: nit: return [image], [image2]?

**Review by BowenBao — COMMENTED** [review]
Could you add more info as for how fold_if pass is failing?

File: torch/onnx/utils.py (hunk 1 above)
Comment (jiafatom) [reply torch/onnx/utils.py:212]: `not bool(dynamic_axes)` means `dynamic_axes == {}`, see [here](https://www.geeksforgeeks.org/python-check-if-dictionary-is-empty/), I prefer this to `not dynamic_axes` to make this explicitly.

**Comment by jiafatom** [comment]
> Could you add more info as for how fold_if pass is failing?

For test_faster_rcnn, when I see the onnx model converted using dummy input, it has only one NonMaxSuppression node,
but when I see the onnx model converted using a true image, it has two NonMaxSuppression nodes.
This is because of the logic that handles differently on some conditions. 
For the existing code, it does not particularly handle dynamic_axes logic, then it wants to treat the input axes as constant and do some folding. Then it gives error when I have a different input shape.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (jiafatom) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:419]: This piece of code is copied from torch vision test, want it to be identical...

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 3 above)
Comment (jiafatom) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:451]: This piece of code is copied from torch vision test, want it to be identical...

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -398,19 +406,7 @@ def run_word_language_model(self, model_name):
         # Only support CPU version, since tracer is not working in GPU RNN.
         self.run_test(model, (x, model.hidden))
 
-    @skipIfUnsupportedOpsetVersion([13])
```
Comment (jiafatom) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:401]: Code move below.

**Comment by KsenijaS** [comment]
dynamic axes are checked on this line of code: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx/fold_if_node.cpp#L54

**Comment by jiafatom** [comment]
> dynamic axes are checked on this line of code: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx/fold_if_node.cpp#L54

Sure. Let's check in this PR first to include the torch vision UT. For the code as mentioned, it breaks current UT, we can fix that in a separate PR.

**Review by neginraoof — APPROVED** [review]
Looks good. Please follow up about the disabled pass. Thanks.

**Pull request merged**